### PR TITLE
added gradient noise optimizer hook

### PR DESCRIPTION
**Pull request by kashif**
This hook function simply adds time-dependent annealed Gaussian noise
to the gradient at every training step. See: [Adding Gradient Noise
Improves Learning for Very Deep Networks](http://arxiv.org/pdf/1511.06807)

### CODE REVIEW

File: chainer/optimizer.py [hunk 1]
```diff
@@ -518,3 +518,52 @@ def __call__(self, opt):
...
+
+    def __call__(self, opt):
+        if cuda.available:
+            kernel = cuda.elementwise(
+                'T noise', 'T g', 'g += noise', 'gradient_noise')
+
+        for param in opt.target.params():
+            p, g = param.data, param.grad
+            xp = cuda.get_array_module(p)
+            noise = xp.random.normal(0,
```
Comment (jheymann85) [inline chainer/optimizer.py]: Thanks for the code. One bug I encountered: This must be moved into the device context (e.g. to line 566), otherwise the hook does not work when using multiple GPUs.

File: chainer/optimizer.py [hunk 2]
```diff
@@ -518,3 +518,51 @@ def __call__(self, opt):
...
+            kernel = cuda.elementwise(
+                'T noise', 'T g', 'g += noise', 'gradient_noise')
+
+        for param in opt.target.params():
+            g = param.grad
+            xp = cuda.get_array_module(g)
+            with cuda.get_device(g) as dev:
+                noise = xp.random.normal(0, numpy.sqrt(
+                    self.eta / numpy.power(1 + opt.t, 0.55)), g.shape).astype(
+                    numpy.float32)
```
Comment (delta2323) [inline chainer/optimizer.py]: Use `g.dtype` instead of `numpy.float32`.

**Comment by delta2323** [comment]
Do you think it is possible to separate the mechanism that determines the quantity of the noise added from the one that does inject the noise to the gradient? Currently these two mechanisms are strongly tied to each other.

**Comment by kashif** [comment]
@delta2323 yes I could do that... but can you kindly let me know where to add the two functions? At the moment the noise is very dependent on the time step of the optimisation, so it made sense to me to confound the two together into this single gradient noise hook function. but if you can tell me where to add the noise helper, then I can try to separate it. Thanks!

**Comment by delta2323** [comment]
One option is to extract the calculation of noise as a function as follows:

```
def exponential_decay_noise(xp, shape, dtype, hook, opt):
    std = numpy.sqrt(hook.eta / numpy.power(1 + opt.t, 0.55))
    return xp.random.normal(0, std, shape).astype(dtype)


class GradientNoise(object):
    def __init__(self, eta, noise_func=exponential_decay_noise):
        self.eta = eta
        self.noise_func = noise_func

    def __call__(self, opt):
        ...
        for param in opt.target.params():
            g = param.grad
            xp = cuda.get_array_module(g)
            with cuda.get_device(g) as dev:
                noise = self.noise_func(xp, g.shape, g.dtype, self, opt)
                ...
```

The function name (`exponential_decay_noise`) and its arguments are not yet considered well. The arguments are the most "informative" ones in the sense that the function takes the hook and optimizer themselves and extract the information that is necessary. I guess `shape` and `dtype` are the most essential arguments.

**Comment by kashif** [comment]
thanks @delta2323 my question was about the noise helper, in which file should it be? would it make sense to put it in `optimizer.py` or somewhere else? thanks!

**Comment by delta2323** [comment]
@kashif Sorry, I forgot to answer the question. I think we put the helper function in `optimizer.py`, or same location as `GradientNoise` hook. We need not to separate it from the hook because we can expect we do not use it other than for this hook.

**Comment by kashif** [comment]
@delta2323 thanks changed as per your suggestion.

File: chainer/optimizer.py [hunk 3]
```diff
@@ -518,3 +523,50 @@ def __call__(self, opt):
...
+
+    . math::
+
+        \\sigma_t^2 = \\frac{\\eta}{(1+t)^\\gamma}
+
+    with :math:`\\eta` selected from {0.01, 0.3, 1.0} and
+    :math:`\\gamma = 0.55`.
+
+    Args:
+        eta (float): parameter for the standard deviation.
```
Comment (delta2323) [inline chainer/optimizer.py]: Please add `noise_func` argument to the description of `Args`.

File: chainer/optimizer.py [hunk 4]
```diff
@@ -518,3 +523,50 @@ def __call__(self, opt):
...
+
+class GradientNoise(object):
+    """Optimizer hook function for adding gradient noise.
+
+    This hook function simply adds time-dependent annealed Gaussian noise
+    to the gradient at every training step:
+
+    . math::
+
+        g_t \\leftarrow g_t + N(0, \\sigma_t^2)
```
Comment (delta2323) [inline chainer/optimizer.py]: As we design this hook so that we can flexibly change the noise function (by `noise_func`), we need to change the document so that this behavior is the default one. We also need to write how to change the noise function as well.

File: chainer/optimizer.py [hunk 5]
```diff
@@ -518,3 +523,50 @@ def __call__(self, opt):
...
+
+    where
+
+    . math::
+
+        \\sigma_t^2 = \\frac{\\eta}{(1+t)^\\gamma}
+
+    with :math:`\\eta` selected from {0.01, 0.3, 1.0} and
```
Comment (delta2323) [inline chainer/optimizer.py]: We do not restrict the range of `eta` to these three values in this hook. So we need to clear that this choice is done in the original paper.

**Comment by delta2323** [comment]
Sorry for my late response. LGTM except comments.

File: chainer/optimizer.py [hunk 6]
```diff
@@ -518,3 +530,52 @@ def __call__(self, opt):
...
+    :math:`\\gamma = 0.55`.
+
+    Args:
+        eta (float): parameter for the noise function, which for the default
```
Comment (delta2323) [inline chainer/optimizer.py]: Could you explain the role of the attribute in more detail? For example, "parameter that defines the scale of the noise".

**Comment by delta2323** [comment]
@kashif Sorry, I have one more comments. Could you check that?

**Comment by delta2323** [comment]
LGTM 👍

**Comment by delta2323** [comment]
Thank you for your contribution!

**Comment by kashif** [comment]
thank you 🙇
